src/main/java/io/lakefs/PathMapper.java

```diff
@@ -0,0 +1,311 @@
+package io.lakefs;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * A class that does path mapping between configured pairs of path styles, and matches the appropriate path mapping for
+ * a converted path.
+ */
+public class PathMapper {
+
+    public static final Logger LOG = LoggerFactory.getLogger(PathMapper.class);
+
+    protected static final String MAPPING_CONFIG_PREFIX = "routerfs.mapping.";
+    protected static final String NON_DEFAULT_MAPPING_CONFIG_PATTERN = "^routerfs\\.mapping\\.(.*)\\.(\\d*)\\.(replace|with)";
+    protected static final String DEFAULT_MAPPING_CONFIG_PATTERN = "^routerfs\\.mapping\\.(.*default)\\.(replace|with)";
+    protected static final int NON_DEFAULT_TO_FS_IDX = 1;
+    protected static final int NON_DEFAULT_MAPPING_IDX_IDX = 2;
+    protected static final int NON_DEFAULT_MAPPING_TYPE_IDX = 3;
+    protected static final int DEFAULT_TO_FS_IDX = 1;
+    protected static final int DEFAULT_MAPPING_TYPE_IDX = 2;
+
+    enum MappingConfigType {
+        REPLACE,
+        WITH
+    }
+
+    private List<PathMapping> pathMappings;
+    private PathMapping defaultMapping;
+
+    public PathMapper(Configuration conf) throws IOException {
+        this.pathMappings = new ArrayList<>();
+        loadMappingConfiguration(conf);
+        if (defaultMapping == null) {
+            throw new MissingDefaultMappingProperty("Missing default mapping configuration, cannot initialize path mapper");
+        }
+    }
+
+    private void loadMappingConfiguration(Configuration conf) throws InvalidPropertiesFormatException {
```

Comment:
You have loadMappingConfiguration and loadMappingConfig - confusing.

Reply:
oops :) fixed, thanks!